cql3/expr/restrictions.cc

```diff
@@ -143,6 +166,19 @@ void preliminary_binop_vaidation_checks(const binary_operator& binop) {
             throw exceptions::invalid_request_exception("LIKE cannot be used for Multi-column relations");
         }
 
+        if(binop.op == oper_t::IN) {
+            auto valid_size = lhs_tup->elements.size();
+            if (is<collection_constructor>(binop.rhs)) {
+                const std::vector<expression>& elements = as<collection_constructor>(binop.rhs).elements;
+                validate_tuples_size (elements, valid_size);                
+            } else if (auto rhs_const = as_if<constant>(&binop.rhs)) {
+                if (rhs_const->type->without_reversed().is_collection()) {
+                    const std::vector<expression>& elements = as<collection_constructor>(*rhs_const).elements;
+                    validate_tuples_size (elements, valid_size);
+                }
+            }
+        }
+
```

Comment:
We should accept assignment with a smaller input tuple than the target.

Example:

```cql
CREATE TABLE q (id int primary key, qq tuple<int, int>);

-- workload
SELECT * FROM q WHERE if = ? AND qq = (?, ?)

-- schema change
ALTER TABLE q ALTER COLUMN qq TYPE tuple<int, int, text>;

-- workload should continue to work
```

We should check whether the rhs is assignable to the lhs, that takes care not only of the tuple sizes, but also of the element types. But doesn't we already do this type check?

Reply:
@avikivity I think the issue you're describing is closer to https://github.com/scylladb/scylladb/issues/12554 than to https://github.com/scylladb/scylladb/issues/13241 which this patch is fixing.

Reply:
They are the same issue. As is comparing tuple<int, text> to tuple<text, int>. If the types are not compatible, they should not be compared. The tuple lengths, whether it's a tuple or not, and the tuple components must all match. This is better done be operating on the type (ask whether the types are compatible), not checking each individual aspect separately.

Reply:
@avikivity , @nyh I think there are two different issues. This issue describes a problem with multi-column clustering keys when right part of expression should have the same length as the left. But @avikivity tells about correspondence of the right part to the schema.
What do you think?